File: binary_tree/105/105_recursion.py
class TreeNode(object):

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution(object):

    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode

        The most important thing is finding the index of
        root in `inorder` according to the value of root
        in `preorder`.
        """
        if len(preorder) == 0:
            return
        val = preorder[0]
        root = TreeNode(val)
        rootIdx = inorder.index(val)
        root.left = self.buildTree(preorder[1:rootIdx + 1], inorder[:rootIdx])
        root.right = self.buildTree(preorder[rootIdx + 1:],
                                    inorder[rootIdx + 1:])
        return root

    def buildTreeHelper(self, preorder, inorder):
        if preorder is None or inorder is None:
            return
        val = preorder[0]
        root = TreeNode(val)
        rootIdx = inorder.index(val)
        root.left = self.buildTreeHelper(preorder[1:rootIdx + 1],
                                         inorder[:rootIdx])
        root.right = self.buildTreeHelper(preorder[rootIdx + 1:],
                                          inorder[rootIdx + 1:])
        return root

File: binary_tree/105/105_recursion.cpp
#include <vector>

using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder) {
        return buildTree(preorder, inorder, 0, 0, (int)preorder.size() - 1);
    }

    /** `preorder` tells us the left and right child root of this subroot,
     *  `inorder` tells us the intervals of left children and right children.
     *  @param root the root index in preorder
     *  @param left the minimum index in inorder that we have to consider
     *  @param right the maximum index in inorder that we have to consider
     */
    TreeNode *buildTree(vector<int> &preorder, vector<int> &inorder, int root, int left, int right) {
        if (root < 0 || root > (int)preorder.size() - 1) {
            return nullptr;
        }
        int val = preorder[root];
        int rootInInorder = indexOf(val, inorder);
        if (rootInInorder < left || rootInInorder > right) {
            return nullptr;
        }
        int leftRoot = root + 1;
        int leftSize = rootInInorder - left;
        int rightRoot = leftRoot + leftSize;
        TreeNode *rootNode = new TreeNode(val);
        rootNode->left = buildTree(preorder, inorder, leftRoot, left, rootInInorder - 1);
        rootNode->right = buildTree(preorder, inorder, rightRoot, rootInInorder + 1, right);
        return rootNode;
    }

    int indexOf(int val, vector<int> &inorder) {
        int i = 0;
        for (; i < (int)inorder.size(); ++i) {
            if (inorder[i] == val) {
                break;
            }
        }
        return i;
    }
};